back/node_watcher/src/index.ts

```diff
@@ -3,17 +3,24 @@
 // of the Apache-2.0 license. See the LICENSE file for details.
 
 import { logger } from '@polkadot/util';
+import pRetry from 'p-retry';
 
 import { nodeWatcher } from './nodeWatcher';
-import { nomidotTasks } from './tasks';
 
 const l = logger('main');
 
-function main(): void {
-  nodeWatcher(nomidotTasks).catch(e => {
-    l.error(e);
-    process.exit(1);
+async function main(): Promise<void> {
+  await pRetry(nodeWatcher, {
```

Comment:
This will start the whole while loop again from `START_FROM || 0`.

I think it would be better to put the pRetry inside the while loop, so that we only retry the last loop

Reply:
Well, actually, for now we can restart the whole thing again, less refactoring.

Reply:
This current implementation is letting the api connection open, we indeed recreate a new api connection at every retry. I can see it simply launching the node_watcher without launching my node:
```bash
$ node -r ts-node/register --max-old-space-size=8192 ./src/index.ts
2020-03-25 14:02:13          API-WS: disconnected from ws://0.0.0.0:9944 code: '1006' reason: 'connection failed' 
# ^--- 1 connection
Attempt 1 failed. There are 10 retries left.
2020-03-25 14:02:14          API-WS: disconnected from ws://0.0.0.0:9944 code: '1006' reason: 'connection failed'
2020-03-25 14:02:14          API-WS: disconnected from ws://0.0.0.0:9944 code: '1006' reason: 'connection failed'
# ^--- 2 connections
Attempt 2 failed. There are 9 retries left.
2020-03-25 14:02:15          API-WS: disconnected from ws://0.0.0.0:9944 code: '1006' reason: 'connection failed'
2020-03-25 14:02:15          API-WS: disconnected from ws://0.0.0.0:9944 code: '1006' reason: 'connection failed'
2020-03-25 14:02:16          API-WS: disconnected from ws://0.0.0.0:9944 code: '1006' reason: 'connection failed' 
# ^--- 3 connections
Attempt 3 failed. There are 8 retries left.
...
```

Reply:
I wouldn't expect this since the `ApiPromise.create` is throwing. Still the api seems to be created. Not sure how I'd be able to disconnect it in the `catch`.. since I can't access the `api` there.

Reply:
What I meant is: with your implementation, if there's an api error, we restart the while loop from the beginning.

But I think it's a good idea as a 1st step.

Reply:
Ok, this I understood, then I'm talking about something else. From my tests, there is definitely something wrong with several instances of the api running, I launched `node_wacher` on a running local node, then stop the node a couple seconds, it retries 2/3 times, and then I relaunch the node. then node watcher prints things like:
```bash
2020-03-25 14:26:21    NODE-WATCHER: Writing: []
2020-03-25 14:26:21    NODE-WATCHER: blockIndex: 23
2020-03-25 14:26:21    NODE-WATCHER: lastKnownBestBlock: 566
2020-03-25 14:26:21    NODE-WATCHER: lastKnownBestFinalized: 564
2020-03-25 14:26:21    NODE-WATCHER: block: 23
2020-03-25 14:26:21    NODE-WATCHER: hash: 0xf5fd811dd1ff793ef74d9a3b779d062a09de514c400bbc0c4fad5b3d57381147
2020-03-25 14:26:21    NODE-WATCHER: blockIndex: 24
2020-03-25 14:26:21    NODE-WATCHER: lastKnownBestBlock: 566
2020-03-25 14:26:21    NODE-WATCHER: lastKnownBestFinalized: 564
2020-03-25 14:26:21    NODE-WATCHER: block: 24
2020-03-25 14:26:21    NODE-WATCHER: hash: 0x73a7d780ee68ba88640a4f61e8617aa863f184172ff43c2afbf517f09fae5e90
2020-03-25 14:26:21    NODE-WATCHER: Task --- createBlockNumber
2020-03-25 14:26:21    NODE-WATCHER: Task --- createBlockNumber
``` 

which shows that several instances are running. This can't be right, but I'm not sure how to prevent that.

Reply:
I've tried a variant by creating the api object outside the retry, and pass it to nodewatcher. The same happen unfortunatly: [here is the branch](https://github.com/paritytech/Nomidot/compare/tbaut-retry...tbaut-1-api?expand=1).

I can't explain why.

Reply:
> which shows that 2 instances are running. This can't be right, but I'm not sure how to prevent that.

I think it's actually correct. Maybe instead of `while(true)`, we can put `while(noError)`, and set `noError=true` at the beginning, then `=false`, whenever we reject.

I'm pretty sure there's a cleaner solution than this, but can't think of one right now.